lib/types/iprange.go

```diff
@@ -0,0 +1,262 @@
+			if n+1 > n {
+				n++
+			}
+			return n
+		default:
+			return (uint64(1) << zbits) - offset
+		}
+	}
+	hostNum, netNum := offsetCIDR(hz, hk-h0), offsetCIDR(nz, nk-n0)
+	// remove head IP like x.x.x.0 from CIDR
+	if hk == h0 && hz > 0 {
+		hk++
+		hostNum--
+	}
+	// remove tail IP like x.x.x.255 from CIDR
+	if hz > 1 {
+		hostNum--
+	}
+	return &IPBlock{
+		ip:        ipk,
+		hostStart: hk,
+		netStart:  nk,
+		hostN:     hostNum,
+		netN:      netNum,
+		weight:    hostNum,
+		split:     hostNum,
+		mode:      roundRobin,
+	}
+}
+
+// GetRandomIP return a random IP by ID from an IP block
+func (b IPBlock) GetRandomIP(id uint64) net.IP {
+	r := hashIDToUint64(id % b.weight)
+	if ip4 := b.ip.To4(); ip4 != nil {
+		i := b.hostStart + r%b.hostN
+		return net.IPv4(byte(i>>24), byte(i>>16), byte(i>>8), byte(i))
+	}
+	if ip6 := b.ip.To16(); ip6 != nil {
+		netN := b.netStart + r%b.netN
+		hostN := b.hostStart + r%b.hostN
+		if hostN < b.hostStart {
+			netN++
+		}
+		if ip := make(net.IP, net.IPv6len); ip != nil {
+			binary.BigEndian.PutUint64(ip[:8], netN)
+			binary.BigEndian.PutUint64(ip[8:], hostN)
+			return ip
+		}
+	}
+	return nil
+}
+
+// GetRoundRobinIP return a IP by indexes from an IP block
+func (b IPBlock) GetRoundRobinIP(hostIndex, netIndex uint64) net.IP {
+	if ip4 := b.ip.To4(); ip4 != nil {
+		i := b.hostStart + hostIndex%b.weight
+		return net.IPv4(byte(i>>24), byte(i>>16), byte(i>>8), byte(i))
+	}
+	if ip6 := b.ip.To16(); ip6 != nil {
+		netN := b.netStart + netIndex%b.netN
+		hostN := b.hostStart + hostIndex%b.weight
+		if hostN < b.hostStart {
+			netN++
+		}
+		if ip := make(net.IP, net.IPv6len); ip != nil {
+			binary.BigEndian.PutUint64(ip[:8], netN)
+			binary.BigEndian.PutUint64(ip[8:], hostN)
+			return ip
+		}
+	}
+	return nil
+}
+
+// GetPool return an IPBlock slice with corret weight and mode
+// Possible format is range1[:mode[:weight]][,range2[:mode[:weight]]]
+func GetPool(ranges string) IPPool {
+	ss := strings.Split(strings.TrimSpace(ranges), ",")
+	pool := make([]IPBlock, 0)
+	for _, bs := range ss {
+		// range | mode | weight
+		rmw := strings.Split(strings.TrimSpace(bs), "|")
+		sz := len(rmw)
+		if sz < 1 {
+			continue
+		}
+		rangeStr := strings.TrimSpace(rmw[0])
+		r := GetIPBlock(rangeStr)
+		if r == nil {
+			continue
+		}
+		r.weight = r.hostN
+		r.mode = roundRobin
+		if sz > 1 {
+			modeStr := strings.TrimSpace(rmw[1])
+			mode, err := strconv.Atoi(modeStr)
+			if err == nil {
+				r.mode = selectMode(mode)
+			}
+		}
+		if sz > 2 {
+			weightStr := strings.TrimSpace(rmw[2])
+			weight, err := strconv.ParseUint(weightStr, 10, 64)
+			if err == nil {
+				r.weight = weight
+			}
+		}
+		r.split = r.weight
+		if len(pool) > 0 {
+			r.split += pool[len(pool)-1].split
+		}
+		pool = append(pool, *r)
+	}
+	return pool
+}
+
+// GetIP return an IP from a pool of IPBlock slice
+func (pool IPPool) GetIP(id uint64) net.IP {
+	if len(pool) < 1 {
+		return nil
+	}
+	idx := id % pool[len(pool)-1].split
```

Comment:
You should just make the pool a struct of (at least) 2 fields:
1. the slice of ipBlock
2. the count of ipBlock hosts in those ipBlocks

Reply:
the host count is actual IPBlock.hostN which will be used for VU_id % hostN to pick a new IP sequentially, so why move it out of IPBlock?